src/Avalonia.Styling/Styling/Styles.cs

```diff
@@ -54,43 +63,53 @@ public IResourceDictionary Resources
             {
                 value = value ?? throw new ArgumentNullException(nameof(Resources));
 
-                var hadResources = false;
-
-                if (_resources != null)
+                if (Owner is object)
                 {
-                    hadResources = _resources.Count > 0;
-                    _resources.ResourcesChanged -= NotifyResourcesChanged;
+                    _resources?.RemoveOwner(Owner);
                 }
 
                 _resources = value;
-                _resources.ResourcesChanged += NotifyResourcesChanged;
 
-                if (hadResources || _resources.Count > 0)
+                if (Owner is object)
                 {
-                    ((ISetResourceParent)this).ParentResourcesChanged(new ResourcesChangedEventArgs());
+                    _resources.AddOwner(Owner);
                 }
             }
         }
 
-        /// <inheritdoc/>
-        IResourceNode? IResourceNode.ResourceParent => _parent;
-
-        /// <inheritdoc/>
         bool ICollection<IStyle>.IsReadOnly => false;
 
-        /// <inheritdoc/>
+        bool IResourceNode.HasResources
+        {
+            get
+            {
+                if (Count > 0)
+                {
+                    return true;
+                }
+
+                foreach (var i in this)
+                {
```

Comment:
We get here only when Count is 0, so wouldn’t this loop be a no-op?

Reply:
Yep, brainfart here. The `Count` check can be removed.

Reply:
Oh, no, actually it was supposed to be checking `_resources.Count`.